### PR TITLE
Adding multi-objective optimizer based on NSGA-II algorithm

### PR DESCRIPTION
**Pull request by vaisaxena**
- Added NSGA2 optimizer under lalegpl/lib/lale/nsga2.py
- Added test case for NSGA2 optimizer under test/ directory
- Added a demo notebook showing multi-objective optimization of LALE pipeline using OptimizeLast and NSGA2
- Updated setup.py to add [playpus](https://github.com/Project-Platypus/Platypus/) package (platypus-opt) as dependency. NSGA2 operator internally uses implementation of NSGA-II algorithm from platypus package. 

Currently, NSGA2 optimizer doesn't support 'max_eval_time' parameter but there is plan to add this going forward (not planned to be added as part of this PR).

### CODE REVIEW

**Comment by vaisaxena** [comment]
Some additional information about the new NSGA2 based multi-objective optimizer proposed to be added through this PR. 

The NSGA2 optimizer supports following parameters:

- **estimator:** LALE Individual Operator.  Its hyper-parameters, obtained through get_param_ranges(), will be tuned for performing MOO. Example: estimator = LGBMClassifier()
- **scoring:**  List of scorers where each scorer is similar to what is supported in Hyperopt. Provide as many scorers as the number of objectives one wants to optimize. Example: scoring = ['roc_auc', make_scorer(compute_fpr, greater_is_better=False)]
- **best_score:** List of best scores for each scorer  specified in scoring parameter above.  Similar to Hyperopt. 
Example: best_score = [1, 0] for example scoring specified above, as 'roc_auc' and FPR scorers have best scores of 1 and 0 respectively. 
- **cv:** Cross Validation parameter similar to Hyperopt supporting integer value or CV object supporting 'split' method. 
            If cv=1, the behavior is similar to passing 'handle_cv_failure=True' in Hyperopt. In this case, the input to fit is divided   
           between train-val splits in 80:20 ratio and the scores are computed on the validation split.   
- **max_evals:** Max number of evaluations when performing MO optimization, similar to Hyperopt.   
-  **max_opt_time:** Maximum time budget for optimization. Similar to Hyperopt. 
- **population_size:** Population size used by NSGA-II algorithm.
- **random_seed:** Only used when cv=1. When cv=1, this parameter is used as _random_state_ in train_test_split() to split the dataset between training and validation splits.

The NSGA2 optimizer supports following routines (similar to Hyperopt):
 - **summary():** Returns list of pareto-optimal pipelines along with the corresponding loss values for the specified scorers. 
 - **get_pipeline(pipeline_name='pn'):** Get the pipeline corresponding to the specified pipeline name obtained from the summary() table. Return the first pipeline from summary() if no pipeline name is specified. 
 - **predict(X, pipeline_name='pn') :** Predict using the pipeline corresponding to the specified pipeline name obtained from the summary() table. Predict using the first pipeline from summary() table if no pipeline name is specified.

**Review by hirzel — APPROVED** [review]
Thank you so much for your PR! It looks great overall. In particular, I am happy that you included several regression tests as well. One remark is that in the notebook demo, the call to `summary()` reveals that the column names are not meaningful. Instead of columns for `loss0`, `loss1` it would be more intuitive to see `-accuracy`, `fpr`. One way to do that would be for the `scoring` argument to accept a dictionary with metric names as keys and scorers as values. That's one of the supported formats of the `scoring` argument of [GridSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html). This is just a suggestion and does not need to hold up the PR.

**Comment by vaisaxena** [comment]
Thanks @hirzel for the comments. 

Regarding the tests, @shinnar in fact suggested me to develop the tests following the same convention as followed in [test_optimizers.py](https://github.com/IBM/lale/blob/master/test/test_optimizers.py)

Your remark about providing the meaningful names to columns in summary() instead of loss1, loss2 etc makes perfect sense. I can change scoring parameter to be a dictionary instead of list with keys as metric names (str) and values as either pre-defined sklearn scorer strings (such as 'accuracy') or a scorer callable object (such as created with make_scorer). So in this case, the earlier specified example of scoring would become
 ```
scoring = {'AUC' : 'roc_auc', 
                   'FPR' : make_scorer(compute_fpr, greater_is_better=False)}
```
and with best_score = [1,0], the column names in the summary() table would become `'best_score[0] - AUC scorer value'` and '`best_score[1] - FPR scorer value`', i.e. `'1-AUC'` and `'FPR' `instead of loss1 and loss2 respectively. Let me know if this is consistent with your remark.  
This is a good suggestion and I will either make this change as part of this PR or a subsequent PR.

**Comment by hirzel** [comment]
Allowing a dictionary from metric names to scorers for the scoring argument sounds like a great approach, given that it is one of the formats that sklearn also supports. Also, I agree that you can make this change as part of this PR or a subsequent PR.

File: lalegpl/lib/lale/nsga2.py [hunk 1]
```diff
@@ -0,0 +1,567 @@
...
+            moo_solutions.append(self.Soln(vars_dict, solution.objectives))
+            logger.info(f"{vars}, {solution.objectives}")
+
+        self.moo_solutions = moo_solutions
+
+        pareto_models = []
+        for solution in self.moo_solutions:
+            est = self.model_helper.create_instance(solution.variables)
+            est_trained = est.fit(X, y)
+            pareto_models.append((solution.variables, est_trained))
```
Comment (rithram) [inline lalegpl/lib/lale/nsga2.py:376]: There might be many many pareto models and training all of them might take up a lot of time post-optimization. It might be worthwhile to limit this number (for example, 10 or 25). Then we can somehow pick at most that many pareto solutions and train them. Picking the pareto solutions might be some additional work (but much cheaper computationally relative to the actual MOO).

File: lalegpl/lib/lale/nsga2.py [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+
+    def get_pareto_models(self):
+        return self.pareto_models
+
+    # Predict using first pareto-optimal estimator
+    def predict(self, X, **kwargs):
+        if "pipeline_name" in kwargs:
+            pname = kwargs["pipeline_name"]
+            pipeline = self.get_pipeline(pipeline_name=pname)
+            del kwargs["pipeline_name"]
```
Comment (rithram) [inline lalegpl/lib/lale/nsga2.py:392]: Why are we deleting something from an object sent by the user? Maybe this is something regarding python that i don't follow.

File: lalegpl/lib/lale/nsga2.py [hunk 3]
```diff
@@ -0,0 +1,567 @@
...
+        astype : 'lale' or 'sklearn', default 'lale'
+            Type of resulting pipeline.
+
+        Returns
+        -------
+        result : Trained operator."""
+
+        id = 0
+        if pipeline_name is not None:
+            id = int(pipeline_name[1:])
```
Comment (rithram) [inline lalegpl/lib/lale/nsga2.py:422]: Minor:
IMO, this appears to be a somewhat non-robust way of getting `id` from `pipeline_name`. Might be worthwhile to have a `(name, pipeline)` dictionary or some guidance to the user as to how the pipeline names should be set.

**Review by rithram — APPROVED** [review]
Thank for the PR. I have left some comments that can be addressed (in the future) or ignored if you think they are not valid or important. Couple of other comments:
- The NB example is really nice. It might be worthwhile have a non-`OptimizeLast` version too with just some operator like `LightGBM`
- The regression tests are quite thorough and covers various aspects. That is very nice!

File: lalegpl/lib/lale/nsga2.py (hunk 2 above)
Comment (hirzel) [reply lalegpl/lib/lale/nsga2.py:392]: Maybe use [dict_without](https://github.com/IBM/lale/blob/90c2348d0394a4cd7baaaeb4546f6589338ad382/lale/helpers.py#L145)?
